include/targets/cuda/math_helper.cuh

```diff
@@ -0,0 +1,124 @@
+  };
+
+  template <> struct rsqrt_impl<true> {
+    template <typename T> __device__ inline T operator()(T a) { return ::rsqrt(a); }
+  };
+
+  /**
+   * @brief Reciprocal square root function (rsqrt)
+   * @param a the argument  (In|out)
+   *
+   * some math libraries provide a fast inverse sqrt() function.
+   * this implementation uses the CUDA builtins
+   */
+  template<typename T> inline __host__ __device__ T rsqrt(T a) { return target::dispatch<rsqrt_impl>(a); }
+
+  /**
+     Generic wrapper for Trig functions -- used in gauge field order
+  */
+  template <bool isFixed, typename T>
+  struct Trig {
+    __device__ __host__ static T Atan2( const T &a, const T &b) { return ::atan2(a,b); }
+    __device__ __host__ static T Sin( const T &a ) { return ::sin(a); }
+    __device__ __host__ static T Cos( const T &a ) { return ::cos(a); }
+    __device__ __host__ static void SinCos(const T &a, T *s, T *c) { sincos(a, s, c); }
+  };
+
+  template <bool is_device> struct sinf_impl { inline float operator()(const float& a) { return ::sinf(a); } };
+  template <> struct sinf_impl<true> { __device__ inline float operator()(const float& a) { return __sinf(a); } };
+
+  template <bool is_device> struct cosf_impl { inline float operator()(const float& a) { return ::cosf(a); } };
+  template <> struct cosf_impl<true> { __device__ inline float operator()(const float& a) { return __cosf(a); } };
+
+  /**
+     Specialization of Trig functions using floats
+   */
+  template <>
+    struct Trig<false,float> {
+    __device__ __host__ static float Atan2( const float &a, const float &b) { return ::atan2f(a,b); }
+    __device__ __host__ static float Sin(const float &a) { return target::dispatch<sinf_impl>(a); }
+    __device__ __host__ static float Cos(const float &a) { return target::dispatch<cosf_impl>(a); }
+    __device__ __host__ static void SinCos(const float &a, float *s, float *c) { target::dispatch<sincosf_impl>(a, s, c); }
+  };
+
+  /**
+     Specialization of Trig functions using fixed b/c gauge reconstructs are -1 -> 1 instead of -Pi -> Pi
+   */
+  template <>
+    struct Trig<true,float> {
+    __device__ __host__ static float Atan2( const float &a, const float &b) { return ::atan2f(a,b) / static_cast<float>(M_PI); }
+    __device__ __host__ static float Sin(const float &a) { return target::dispatch<sinf_impl>(a * static_cast<float>(M_PI)); }
+    __device__ __host__ static float Cos(const float &a) { return target::dispatch<cosf_impl>(a * static_cast<float>(M_PI)); }
+    __device__ __host__ static void SinCos(const float &a, float *s, float *c) { target::dispatch<sincosf_impl>(a * static_cast<float>(M_PI), s, c); }
+  };
+
+  template <bool is_device> struct fpow_impl { template <typename real> inline real operator()(real a, int b) { return std::pow(a, b); } };
+
+  template <> struct fpow_impl<true> {
+    template <typename real> __device__ inline real operator()(real a, int b)
+    {
+      if (sizeof(real) == sizeof(double)) {
```

Comment:
I think it'd be safer to have a default case of `::pow(a,b)`, then have a specialization for `float`---I think we'd have to use SFINAE (`::is_float`?) to get around normal issues with partial template specialization?

As the code is written, I think it's technically legal to pass in a `double2` (`==complex<double>`). That might give a compiler error technically, but better to be safe than sorry.